### PR TITLE
Enable Windows e2e logging

### LINKED ISSUE #1286: Log dumping script for Windows clusters
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Our log dumping script (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/hack/log/log-dump.sh) should support Linux as well as Windows clusters 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: N/A
- Kubernetes version: (use `kubectl version`): N/A
- OS (e.g. from `/etc/os-release`): N/A

### PR DESCRIPTION
**Pull request by jsturtevant**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
We can't diagnose windows CI failures with out the logs from the workload clusters.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes  #1286

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Enable Windows workload log collection in e2e tests
```

### CODE REVIEW

**Comment by jsturtevant** [comment]
This currently has the patches from #1350. I will re-base and remove once that is merged. 

Most of the logs are collected now. There are three logs that are failing for some reason but wanted to run and see if we can capture the error that is occurring during Windows e2e.

**Comment by jsturtevant** [comment]
looks like the node didn't come online at all
/test pull-cluster-api-provider-azure-e2e-windows

**Comment by jsturtevant** [comment]
ran into #1286 and got logs from the windows node: https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1351/pull-cluster-api-provider-azure-e2e-windows/1387946030623887360/artifacts/clusters/capz-e2e-vjj1yu/machines/capz-e2e-vjj1yu-md-win-76d8f65f89-fs7tg/

looking to the failure

**Comment by jsturtevant** [comment]
I also found that logs are not collected for machinepools: #1352

**Comment by CecileRobertMichon** [comment]
Looks like this still has patches from #1350 ?

**Comment by jsturtevant** [comment]
> Looks like this still has patches from #1350 ?
yes, added them back for testing.

**Comment by jsturtevant** [comment]
Running Windows job again to see if restart of kubelet works

/test pull-cluster-api-provider-azure-e2e-windows

**Comment by jsturtevant** [comment]
try #2 passed, try 3:

/test pull-cluster-api-provider-azure-e2e-windows

**Comment by jsturtevant** [comment]
try 3 passed, try 4:

/test pull-cluster-api-provider-azure-e2e-windows

**Comment by jsturtevant** [comment]
try 4 failed with nodes failing to come online, logged: #1359

try 5:

/test pull-cluster-api-provider-azure-e2e-windows

**Comment by jsturtevant** [comment]
try 5 passed, try 6:

/test pull-cluster-api-provider-azure-e2e-windows

**Comment by jsturtevant** [comment]
try 6 passed, I've re-based to remove the patches from  #1350

**Comment by CecileRobertMichon** [comment]
If this is actually fixing stuff (besides logging), let's add that in the pr description / title

**Comment by jsturtevant** [comment]
> If this is actually fixing stuff (besides logging), let's add that in the pr description / title

Split out the fix to #1366

File: test/e2e/azure_logcollector.go [hunk 1]
```diff
@@ -20,12 +20,16 @@ package e2e
...
 	"context"
+	"fmt"
 	"io/ioutil"
 	"net/http"
 	"path/filepath"
-	"sigs.k8s.io/cluster-api-provider-azure/azure"
 	"strings"
 
+	"sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4"
+
```
Comment (chewong) [inline test/e2e/azure_logcollector.go:30]: nit: remove extra new line

File: scripts/ci-e2e.sh [hunk 2]
```diff
@@ -78,12 +78,14 @@ export KIND_EXPERIMENTAL_DOCKER_NETWORK="bridge"
...
 AZURE_SSH_PUBLIC_KEY_FILE=${AZURE_SSH_PUBLIC_KEY_FILE:-""}
 if [ -z "${AZURE_SSH_PUBLIC_KEY_FILE}" ]; then
+    echo "generating sshkey for e2e"
     SSH_KEY_FILE=.sshkey
     rm -f "${SSH_KEY_FILE}" 2>/dev/null
     ssh-keygen -t rsa -b 2048 -f "${SSH_KEY_FILE}" -N '' 1>/dev/null
     AZURE_SSH_PUBLIC_KEY_FILE="${SSH_KEY_FILE}.pub"
 fi
 export AZURE_SSH_PUBLIC_KEY_B64=$(cat "${AZURE_SSH_PUBLIC_KEY_FILE}" | base64 | tr -d '\r\n')
+export AZURE_SSH_PUBLIC_KEY=$(cat "${AZURE_SSH_PUBLIC_KEY_FILE}" | tr -d '\r\n')
```
Comment (chewong) [inline scripts/ci-e2e.sh:88]: Could you add a comment on where this is being used?

File: test/e2e/azure_logcollector.go [hunk 3]
```diff
@@ -62,9 +65,27 @@ func collectLogsFromNode(ctx context.Context, managementClusterClient client.Cli
...
+		// due to Windows name limitations in Azure so use ipaddress instead
+		if len(m.Status.Addresses) > 0 {
+			hostname = m.Status.Addresses[0].Address
+		} else {
+			fmt.Printf("INFO: Unable to collect logs as node doesn't have addresses")
+		}
+	}
+
 	port := e2eConfig.GetVariable(VMSSHPort)
+
```
Comment (chewong) [inline test/e2e/azure_logcollector.go:88]: ditto

File: test/e2e/azure_logcollector.go [hunk 4]
```diff
@@ -78,7 +99,37 @@ func collectLogsFromNode(ctx context.Context, managementClusterClient client.Cli
...
+	azMachine := &v1alpha4.AzureMachine{}
+	if err := managementClusterClient.Get(ctx, key, azMachine); err != nil {
+		return false, err
+	}
+
+	if azMachine.Spec.OSDisk.OSType == azure.WindowsOS {
+		return true, nil
+	}
+	return false, nil
```
Comment (chewong) [inline test/e2e/azure_logcollector.go:128]:
```suggestion
	return azMachine.Spec.OSDisk.OSType == azure.WindowsOS, nil
```

File: test/e2e/azure_logcollector.go [hunk 5]
```diff
@@ -78,7 +99,37 @@ func collectLogsFromNode(ctx context.Context, managementClusterClient client.Cli
...
+	if isWindows {
+		// if we initiate to many ssh connections they get dropped (default is 10) so split it up
+		var errors []error
+		errors = append(errors, kinderrors.AggregateConcurrent(windowsInfo(execToPathFn)))
+		errors = append(errors, kinderrors.AggregateConcurrent(windowsK8sLogs(execToPathFn)))
+		errors = append(errors, kinderrors.AggregateConcurrent(windowsNetworkLogs(execToPathFn)))
+		return kinderrors.NewAggregate(errors)
+	} else {
+		return kinderrors.AggregateConcurrent(linuxLogs(execToPathFn))
+	}
```
Comment (chewong) [inline test/e2e/azure_logcollector.go:111]:
```suggestion
	} 
	return kinderrors.AggregateConcurrent(linuxLogs(execToPathFn))
```

**Review by chewong — COMMENTED** [review]
Few nits, otherwise LGTM

File: test/e2e/azure_logcollector.go [hunk 6]
```diff
@@ -62,9 +65,27 @@ func collectLogsFromNode(ctx context.Context, managementClusterClient client.Cli
 	if err != nil {
 		return err
 	}
+
+	fmt.Printf("INFO: Collecting logs for machine %s in cluster %s in namespace %s\n", m.GetName(), cluster.Name, cluster.Namespace)
```
Comment (devigned) [inline test/e2e/azure_logcollector.go:69]: Perhaps, use https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/9d34960d60f48a8b4a1723af9cecd747d7885aa2/test/e2e/log.go#L37

**Review by devigned — CHANGES_REQUESTED** [review]
Little fix up for logging. 

Outside of that, it would be super cool to see the log collection functionality end up in https://github.com/Azure/azure-capi-cli-extension. It would be so useful to be able to ask a user to gather logs from their cluster for debugging.

File: test/e2e/azure_logcollector.go (hunk 1 above)
Comment (jsturtevant) [reply test/e2e/azure_logcollector.go:30]: interesting goimports didn't fix this?  I also thought we ran go imports in the validation scripts

**Comment by nader-ziada** [comment]
/lgtm

over to @devigned

**Review by devigned — APPROVED** [review]
/lgtm
/approve

**Pull request merged**